src/lib/rewrites.ml

```diff
@@ -4186,6 +4186,147 @@ let rewrite_truncate_hex_literals _type_env defs =
+            , atyp                    (* 'increasing' in our example *)
+            , e_loop_body             (* 'f(k, foo, bar) + k'  in our example *)
+            ) ->
+      (
+
+        (* We get the int values of the bounds from their types inferred by the typer *)
+        let int_of_tannot_opt tannot =
+          let int_of_typ_aux_opt : typ_aux -> Z.t option = function
+            | Typ_app (id, [A_aux (A_nexp nexp, _)]) when Id.compare id (mk_id "atom") = 0 ->
+              int_of_nexp_opt @@ nexp_simp nexp
+            | _ -> None
+          in
+          match destruct_tannot tannot with
+          | Some (_, Typ_aux (typ, l)) -> int_of_typ_aux_opt typ
+          | None -> None
+        in
+        let n_start_opt = int_of_tannot_opt tannot1 in
+        let n_end_opt   = int_of_tannot_opt tannot2 in
+        let n_step_opt  = int_of_tannot_opt tannot3 in
+
+        (* Abort unrolling with a warning when types infered are too complex (i.e. not immediate literals) *)
+        match n_start_opt, n_end_opt, n_step_opt with
+        | Some n_start, Some n_end, Some n_step -> (
+
+            (* Build a range out of the bounds
+               in our example, from (start=0, end=3, step=1)
+               the range will be the list [0; 1; 2; 3]
+            *)
+            let range = list_of_ord_range atyp n_start n_end n_step in
+
+            (* Only unroll "small" loops, i.e. those with less than 'max_iter' iterations *)
+            if !opt_unroll_loops_max_iter <> 0
+               && List.length range > !opt_unroll_loops_max_iter
+            then E_aux(e, annot)
+            else (
+
+                (* Build the final expression, a block of n times the body *)
+                let bodies = List.map
+                    (fun z -> rewrite_exp_replace_id_with_num "i" z e_loop_body)
+                    range
+                in
+                E_aux (E_block bodies, annot)
+            ) (* else *)
+          ) (* Some n_start... *)
+
+        | _ -> (
+            let e' = E_aux (e, annot) in
+            let (l : Parse_ast.l), _tannot = annot in
+            let string_of_position (p : Lexing.position) : string =
+              p.pos_fname ^ ":" ^ (string_of_int p.pos_lnum)
+            in
+            let rec string_of_l : Parse_ast.l -> string = function
+            | Unknown                -> "Unknown"
+            | Unique (i, l)          -> Printf.sprintf "Unique( %s, %s)" (string_of_int i) (string_of_l l)
+            | Generated l            -> Printf.sprintf "Generated( %s ) " (string_of_l l)
+            | Hint (str, l1, l2)     -> Printf.sprintf "Hint( %s, %s, %s )" str (string_of_l l1) (string_of_l l2)
+            | Range (p1, p2)         -> Printf.sprintf "Range( %s, %s )" (string_of_position p1) (string_of_position p2)
+            in
+            print_endline  "[WARNING] Cannot unroll the loop because the bounds numerical values couldn't be fully determined.";
```

Comment:
Should try to use `Reporting.warn`, although I think it might be the case that warnings are be supressed during the rewrites to avoid spurious warning now.

Reply:
Oh I didn't know about `Reporting`, that's nice, thank you !

However the warning is not shown, even with `--all-warnings` enabled, should we add something like a `?force_show` option to the `warn` function to override the warning suppression ? Or a neater idea ?

Reply:
FYI I tried the `?force_show` thingy here and it works well : https://github.com/rems-project/sail/pull/744/commits/db921e982e189041ce57b75a2fee4040ab919521